### PR TITLE
Add unconditional import when type checking

### PR DESCRIPTION
**Pull request by kandluis**
## What does this PR do?

Without this change, when type checking using `pyre` (and possibly as well using `mypy` though I've not verified this) imports such as:

```
from pytorch_lightning import Trainer
```
raise errors due to the fact that they're only conditionally imported at the module level.

A workaround is to change imports to the following:

```
from pytorch_lightning.trainer import Trainer
```

However, large chunks of our documentation do not reflect this pattern. As such, this small update to the logic in the `__init__.py` function  makes it so that when type checking (`typing.TYPE_CHECKING === True`) the imports are unconditional, fixing any warnings from the checkers.

This is okay since the only time the imports do not occur is when were' in the build-phase of the library.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/__init__.py [hunk 1]
```diff
@@ -76,6 +67,33 @@
...
+else:
+    try:
+        # This variable is injected in the __builtins__ by the build
+        # process. It used to enable importing subpackages of skimage when
+        # the binaries are not built
+        _ = None if __LIGHTNING_SETUP__ else None
+    except NameError:
+        __LIGHTNING_SETUP__: bool = False
+
+    if __LIGHTNING_SETUP__:  # pragma: no-cover
```
Comment (Borda) [inline pytorch_lightning/__init__.py:79]:
```suggestion
    if not typing.TYPE_CHECKING or __LIGHTNING_SETUP__:  # pragma: no-cover
```
can we write it this way just to not maintain double imports?

Comment (kandluis) [reply pytorch_lightning/__init__.py:79]: Haha! Yeah, had the same intuition as you here :( Sadly, it seems the "or" logic is a bit too complex for the type-checker, so it doesn't work.

My understanding is that `typing.TYPE_CHECKING` actually just gets statically replaced with the literal `False` or `True` before type-checking happens. So from the type-checker's point of view, the the suggested change becomes:

```
if True or __LIGHTNING_SETUP__:
   ...
```
so the imports are still conditional.

Comment (kandluis) [reply pytorch_lightning/__init__.py:79]: Ah, yes, here's the explanation of how `TYPE_CHECKING` works: https://github.com/python/typing/issues/230

**Review by ananthsub — APPROVED** [review]
thanks for the fix!

File: pytorch_lightning/__init__.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/__init__.py:79]: I am sorry I could see it in the linked typing issue

File: pytorch_lightning/__init__.py [hunk 2]
```diff
@@ -76,6 +67,33 @@
...
         'metrics',
     ]
+else:
+    try:
+        # This variable is injected in the __builtins__ by the build
+        # process. It used to enable importing subpackages of skimage when
+        # the binaries are not built
+        _ = None if __LIGHTNING_SETUP__ else None
+    except NameError:
+        __LIGHTNING_SETUP__: bool = False
```
Comment (Borda) [inline pytorch_lightning/__init__.py:77]: does this has to be in the else branch?

**Comment by kandluis** [comment]
Accidentally closed. Reopening. Refactored to keep only imports in the if/else branches.

File: pytorch_lightning/__init__.py (hunk 2 above)
Comment (kandluis) [reply pytorch_lightning/__init__.py]: It does not - moved it out ! Good catch :)

**Review by SeanNaren — APPROVED** [review]
I'll approve however not too well versed here, pinging @carmocca @kaushikb11 who may have more input.

**Comment by tchaton** [comment]
@Borda  Any updates ?

**Comment by Borda** [comment]
> @Borda Any updates ?

it is ready to land #6590 just need your confirmation...

**Comment by kandluis** [comment]
Thanks @Borda for merging into refactors! :)

**Pull request closed without merging**